Cloudflare analytics snapshot blog post (#36)
json-schema-org/blog: pages/posts/website-analytics-snapshot-2023.md

---
title: "50+ million requests in 7 days on json-schema.org"
date: 2023-07-31T06:00:00+00:00
type: engineering
cover: /img/posts/2023/website-analytics-snapshot-2023/cover.webp
authors:
  - name: Ben Hutton
    photo: /img/avatars/benhutton.webp
    link: https://twitter.com/relequestual
excerpt: "See what we discovered when looking at over 50 million requests to the JSON Schema org domain from the last 7 days"
---

> I grant you, 7 days isn't an ideal length of time to do an analysis over, but this is what we have for now. Previously, we had basic website analytics, but that only tells a fraction of the story. We run the website on GitHub pages with Cloudflare infront, which provides 7 days of analytics, critically, including direct access traffic.

Reviewer: ```suggestion
I grant you, 7 days isn't an ideal length of time to do an analysis over, but this is what we have for now. Previously, we had basic website analytics, but that only tells a fraction of the story. We run the website on GitHub pages with Cloudflare in front, which provides 7 days of analytics, critically, including direct access traffic.
```

Why now? In addition to providing meta-schemas at versioned URLs, the web site has, and continues to, provide the latest meta-schema at `/schema`. We currently advise people not to be accessing this! We put in a delayed redirect with a message warning people to avoid it, but realistically, we expect most access will be via code and not in a browser.

Maybe now we can remove hosting that URL? For reasons out of the scope of this article, using it really isn't a great idea, and will probably produce results you don't expect. As mentioned, web analytics won't tell us what's happening outside of a browser, so we needed something else… We were already using Cloudflare, so decided to submit for the open source free pro tier plan… while we wait, the cost isn't huge, so I decided we should look at the data now.

> The data in this article is a 7 day snapshot taken at 2023-07-21.

Reviewer 2: ```suggestion
The data in this article is a seven-day snapshot taken at 2023-07-21.
```

Writing style guide (doesn't actually exist, but [example](https://www.scribendi.com/academy/articles/when_to_spell_out_numbers_in_writing.en.html)) #54 - spell out numbers if they're ten or less.

(At the end of the article, I'll share an image from this snapshot, and one from 2023-07-31 for comparison.)

> 

Reviewer: Maybe applying some markdown styling to the caveats list will improve readability.

## What are we sharing here?

It's best if we start with some splashy numbers, then go into what is exactly being accessed and why that's interesting, and finally just some interesting discoveries.

### The big numbers

51.92 million requests

211.47GB of data transfer

1.05 million page views

1.01 million visits

### All the meta-schemas!

About 51 million of those requests are made from something that doesn't declare a browser, client, or bot name. We see Chrome and Firefox take the top spots, and lower down we see CURL and various indexing bots.

In terms of what paths are accessed, and therefore what meta-schemas, I found the results interesting and insightful.

The top result, at 11.82 million, is `/draft-07/schema`. This suggests that draft-07 JSON Schema is the most popular when it comes to people actually using them or authoring them today.

<figure className="mt-10">
  <img className="mx-auto" src="/blog/img/posts/2023/website-analytics-snapshot-2023/Screenshot 2023-07-21 at 12.12.31.webp"/>
  <figcaption className="mt-2 mb-10 text-sm text-center text-gray-500">Web traffic snapshot 2023-07-21."</figcaption>
</figure>

> Draft-04 doesn't appear till number 4 in the list. Rank number 2 and 3 are taken by Draft 2019-09, but not as we might expect… the files there, `links` and `hyper-schema` are for Hyper Schema, as opposed to the primary validation use case of JSON Schema.

Reviewer: ```suggestion
Draft-04 doesn't appear till number 4 in the list. Ranks number 2 and 3 are taken by Draft 2019-09, but not as we might expect… the files there, `links` and `hyper-schema` are for Hyper Schema, as opposed to the primary validation use case of JSON Schema.
```

Where are these requests for Hyper Schema meta-schemas coming from? While the physical location is uninteresting, we can look at the user agent provided. The top two, which covers 7.5 million requests, doesn't really help… `python-requests`. At most, we can see that version 2.27 and 2.28 are roughly in equal use. After that, we jump down to 635k requests, from something identifying as `neptune-client`. Curiosity piqued!

They call it an "Experiment tracking tool and model registry". This is to do with Machine Learning (ML) Ops (MLOps), built for data scientists and ML engineers. I know very little about ML, so I asked my colleague to share his take on the Neptune.ai service, who knows vastly more about this space…

"Piece of software that, when you tweak something about a machine learning model, let's you run some calculations that tell you how your changes affect how good the model is" - [Julian Berman](https://mastodon.social/@JulianWasTaken)

> Well, that's dumbed down enough for me! Thanks Julian!

Reviewer 3: 😂

When you look at the adjacent data, seeing over half of the remaining top 15 user agents, it's pretty clear that most of these requests are coming from IDEs and code editors. We see InteliJ IDEA, WebStorm, and multiple versions of each. But, what we don't see is VSCode.

Sure enough, VSCode, as far as we can tell, [doesn't provide a user agent](https://github.com/microsoft/vscode/issues/188398). At least, when making requests related to JSON Schema. While we love VSCode and the baked-in support for JSON Schema, there are a number of things we wish were a little better, and this is one of them.

Visual Studio provides a user agent (and we'll get to that analysis later), but given Visual Studio Code does not, we can't differentiate between requests it is making, and any others which aren't providing a user agent. (We generally see a user agent is given.)

## Who is accessing draft-04?

The total number of requests to the draft-04 meta-schema is 7.21 million in the last 7 days.

> The top physical locations showed one surprise, with Peru taking the top spot, over the UK and Russia. Otherwise, the USA takes top spot with 5.73 million of those 7.21 million requests.

Reviewer 2: This is a bit redundant with "top" being present three times.

```suggestion
The physical locations showed one surprise, with Peru taking the lead over the UK and Russia. Otherwise, the USA takes top spot with 5.73 million of those 7.21 million requests.
```

It's also a bit confusing.  Is the order Peru, UK, Russia, USA?  If so, is USA really taking a "top spot"?

Author: You're right. It would have made more senes without the comma, but I agree it's not clear.

Reviewer 2: It still feels clunky calling out USA as the "top spot" when you've just mentioned three countries that are above it.

Maybe

> The USA follows them up with...

Author: That's a fair comment. I'll re-work

> This time, the user agent data is much more useful. We don't see the empty agent string till slot number 10, accounting for only 78k requests. In the top spot, we see Java! I'm making an educated guess here, and saying this is related to the Java implementation by Francis Galiegue. They were one of the early core team members of JSON Schema, and created a Java implementation, which hasn't upgraded support since draft-04.

Reviewer 2: "top spot" again

Author: Now it's not duplicated, are you happier with the phrase being used?

## Latest three JSON Schema versions?

Let's look at accessing paths with `draft-07`, `2019-09`, and `2020-12`.

For draft-07, the path `/draft-07/schema` was accessed 11.66 million times. The next largest, at 1.8k, is the release notes… as in a HTML file on the website (oh, look, there's a website there too!).

For 2019-09, we see 20.87 million requests. Why the jump? Both `links` and `hyper-schema` received over 8 million requests each. As mentioned previously, it's really hard to attribute those requests, but let's see what happens when we exclude them. Now we see 4.05 million… not bad! Although, we have to consider, with 2019-09 we broke up the meta-schema into multiple files, specifically the individual vocabularies. We see just above 500k requests for each of the vocabularies, with slightly more for the general purposes meta-schema, and core.

The user agents for 2019-09, with hyper-schema URLs filtered out, are predominantly Java.

What do we expect to see for 2020-12 then? I was curious too.

We see 7.42 million, and none of those are related to Hyper Schema.

> (For reasons I won't go into, we have a meta-schema for Hyper Schema in 2019-09, but not 2020-12.) 1.53 million to `.../meta/core`, and next palace, the validation vocabulary meta-schema, with only 577.84k. If this is accurate, it's a little concerning. We would expect `/draft/2020-12/schema` to be the top path accessed, but it doesn't appear till position 8, with 472.46k.

Reviewer 2: ```suggestion
(For reasons I won't go into, we have a meta-schema for Hyper Schema in 2019-09, but not 2020-12.) 1.53 million to `.../meta/core`, and in the next place, the validation vocabulary meta-schema, with only 577.84k. If this is accurate, it's a little concerning. We would expect `/draft/2020-12/schema` to be the top path accessed, but it doesn't appear till position 8, with 472.46k.
```

Point is, we can tell without needing to look up IP address owners, that almost 80% of our traffic is coming from AWS!

## Country

Given how much traffic is coming from AWS, it's not a surprise that about 43 million of those requests are coming from the US and Ireland. After those, Finland, India, and Netherlands make up the 1 mil club. Following on, Germany and Russia, and then several countries from Asia, all above 200k.

## Source

A staggering 51.12 million of those 51.92 million requests, are not declaring a reference. The following top three are json-schema.org, Google, and then, very interestingly, localhost!

Weirdly, 1.54k requests declare a referrer of some defunct Uber settlement, just above Duck Duck Go.

## What does this all mean?

> Access of meta-schemas is unlikely to corrolate with useage percentages, but it is at least an indicator of use to an extent.

Reviewer: ```suggestion
Access of meta-schemas is unlikely to correlate with useage percentages, but it is at least an indicator of use to an extent.
```

Looking at this data has meant we are reaching out to AWS to see if we can get any details.
It has prompted us to reach out to IDE makers, although doing so was already on our list.
We have extended our known platform users. It's nice when people use a proper agent string.

Given we only have access to a 7 day history, there's not much we can tell over a longer time.
While traditional analytics are useful, and we do use them, they can't tell us about accessing JSON files directly.

> While some of these numbers are interesting to see, I'm yet to be convinced that they provide much utility today.

Reviewer: Considering that we don't have full clarity on the reasons of the traffic I'll suggest to end the article with a call to action to readers to provide feedback or ideas on the reasons of the data or maybe how to perform a deeper analysis. Do we have a github issue to continue the discussion there?

Author: Added. Let me know if you think this is OK by resolving this conversation and leaving an updated review? =]

---
title: "50+ million requests in 7 days on json-schema.org"
date: 2023-07-31T06:00:00+00:00
type: engineering
cover: /img/posts/2023/website-analytics-snapshot-2023/cover.webp
authors:
  - name: Ben Hutton
    photo: /img/avatars/benhutton.webp
    link: https://twitter.com/relequestual
excerpt: "See what we discovered when looking at over 50 million requests to the JSON Schema org domain from the last 7 days"
---

I grant you, 7 days isn't an ideal length of time to do an analysis over, but this is what we have for now. Previously, we had basic website analytics, but that only tells a fraction of the story. We run the website on GitHub pages with Cloudflare in front, which provides 7 days of analytics, critically, including direct access traffic.

The JSON Schema website hosts the official meta-schemas. If you're not sure what that means, meta-schemas are JSON Schemas for JSON Schema. Why is this interesting? Well, we host all historical versions of the meta-schema, not just the most recent. And while we don't expect code to be making requests on a regular basis to download the meta-schemas, the number of requests strongly suggests that they do.

> Why now? In addition to providing meta-schemas at versioned URLs, the web site has, and continues to, provide the latest meta-schema at `/schema`. We currently advise people not to be accessing this! We put in a delayed redirect with a message warning people to avoid it, but realistically, we expect most access will be via code and not in a browser.

Author: Let me know if you find this an acceptable resolution.
https://github.com/json-schema-org/blog/pull/36/commits/fab19eb3b5e484ac9f86d7e00a2661988a95b706

- **Caveat 3**: The "top" of each category of data is limited to 15 items in the UI. I think we could use their API to access additional data, but that's something I couldn't figure out in 5 minutes, so let's use what we have today.

Oh yes, and we are limited to the last SEVEN days worth of data.

## What are we sharing here?

It's best if we start with some splashy numbers, then go into what is exactly being accessed and why that's interesting, and finally just some interesting discoveries.

### The big numbers

51.92 million requests

211.47GB of data transfer

1.05 million page views

1.01 million visits

### All the meta-schemas!

About 51 million of those requests are made from something that doesn't declare a browser, client, or bot name. We see Chrome and Firefox take the top spots, and lower down we see CURL and various indexing bots.

In terms of what paths are accessed, and therefore what meta-schemas, I found the results interesting and insightful.

> The top result, at 11.82 million, is `/draft-07/schema`. This suggests that draft-07 JSON Schema is the most popular when it comes to people actually using them or authoring them today.

Author: That's a good point. I'll think about how to change this.

> The top result, at 11.82 million, is `/draft-07/schema`. This suggests that draft-07 JSON Schema is the most popular when it comes to people actually using them or authoring them today.

Author: I've modified the wording. Let me know what you think.

When you look at the adjacent data, seeing over half of the remaining top 15 user agents, it's pretty clear that most of these requests are coming from IDEs and code editors. We see InteliJ IDEA, WebStorm, and multiple versions of each. But, what we don't see is VSCode.

Sure enough, VSCode, as far as we can tell, [doesn't provide a user agent](https://github.com/microsoft/vscode/issues/188398). At least, when making requests related to JSON Schema. While we love VSCode and the baked-in support for JSON Schema, there are a number of things we wish were a little better, and this is one of them.

Visual Studio provides a user agent (and we'll get to that analysis later), but given Visual Studio Code does not, we can't differentiate between requests it is making, and any others which aren't providing a user agent. (We generally see a user agent is given.)

## Who is accessing draft-04?

The total number of requests to the draft-04 meta-schema is 7.21 million in the last 7 days.

The physical locations showed one surprise, with Peru taking the lead over the UK and Russia. Next the USA with 5.73 million of those 7.21 million requests.

> This time, the user agent data is much more useful. We don't see the empty agent string till slot number 10, accounting for only 78k requests. At the top we see Java! I'm making an educated guess here, and saying this is related to the Java implementation by Francis Galiegue. They were one of the early core team members of JSON Schema, and created a Java implementation, which hasn't upgraded support since draft-04.

Author: I'm talking about this one: https://github.com/java-json-tools/json-schema-validator
There's nothing in the readme.
Were you thinking of a different one?

> This time, the user agent data is much more useful. We don't see the empty agent string till slot number 10, accounting for only 78k requests. At the top we see Java! I'm making an educated guess here, and saying this is related to the Java implementation by Francis Galiegue. They were one of the early core team members of JSON Schema, and created a Java implementation, which hasn't upgraded support since draft-04.

Reviewer 4: I could have sworn that they had added support for some, but not all of draft-07. Looking now, I don't see any evidence that that's the case. I must have conflated that library with something else.

## Latest three JSON Schema versions?

Let's look at accessing paths with `draft-07`, `2019-09`, and `2020-12`.

For draft-07, the path `/draft-07/schema` was accessed 11.66 million times. The next largest, at 1.8k, is the release notes… as in a HTML file on the website (oh, look, there's a website there too!).

For 2019-09, we see 20.87 million requests. Why the jump? Both `links` and `hyper-schema` received over 8 million requests each. As mentioned previously, it's really hard to attribute those requests, but let's see what happens when we exclude them. Now we see 4.05 million… not bad! Although, we have to consider, with 2019-09 we broke up the meta-schema into multiple files, specifically the individual vocabularies. We see just above 500k requests for each of the vocabularies, with slightly more for the general purposes meta-schema, and core.

The user agents for 2019-09, with hyper-schema URLs filtered out, are predominantly Java.

What do we expect to see for 2020-12 then? I was curious too.

We see 7.42 million, and none of those are related to Hyper Schema.

> (For reasons I won't go into, we have a meta-schema for Hyper Schema in 2019-09, but not 2020-12.) 1.53 million to `.../meta/core`, and in the next place, the validation vocabulary meta-schema, with only 577.84k. If this is accurate, it's a little concerning. We would expect `/draft/2020-12/schema` to be the top path accessed, but it doesn't appear till position 8, with 472.46k.

Author: I better fix that then! Thanks for highlighting!

> (For reasons I won't go into, we have a meta-schema for Hyper Schema in 2019-09, but not 2020-12.) 1.53 million to `.../meta/core`, and in the next place, the validation vocabulary meta-schema, with only 577.84k. If this is accurate, it's a little concerning. We would expect `/draft/2020-12/schema` to be the top path accessed, but it doesn't appear till position 8, with 472.46k.

Author: Fixed. Thanks
